### ISSUE TITLE
Enable merge queue in github to avoid commit confliction.

### ISSUE THREAD

**Issue by liurenjie1024** [apache/datafusion#6880]
### Describe the bug

See https://github.com/apache/arrow-datafusion/pull/6875 for details. We can enable merge queue to avoid such kind of conflicts. You can find out details here:
https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-a-merge-queue

**Comment by alamb** [comment]
This would be awesome to do -- thank you for the suggestion @liurenjie1024 

The repository settings are controlled by https://github.com/apache/arrow-datafusion/blob/main/.asf.yaml

Perhaps someone can research how to change that file to enable the merge queue

**Comment by alamb** [comment]
Here is the page: [INFRA-Git-.asf.yamlfeatures-120723-1012-2450.pdf](https://github.com/apache/arrow-datafusion/files/12026638/INFRA-Git-.asf.yamlfeatures-120723-1012-2450.pdf)

**Comment by alamb** [comment]
😢

**Comment by crepererum** [comment]
I propose we re-open this since this is clearly a feature that we want, it's just blocked by ASF at the moment.

**Comment by crepererum** [comment]
The docs and the relevant management code are now openly available:

https://github.com/apache/infrastructure-asfyaml

So whoever wants to make this a reality for DataFusion can start there.

**Comment by blaginin** [comment]
Based on ASF Slack, I believe MQ aren't currently supported in `.asf.yaml` because there's no API support (https://github.com/orgs/community/discussions/50893)

**Comment by crepererum** [comment]
> Based on ASF Slack, I believe MQ aren't currently supported in `.asf.yaml` because there's no API support ([github.com/orgs/community/discussions/50893](https://github.com/orgs/community/discussions/50893))

Following https://github.com/orgs/community/discussions/50893#discussioncomment-7792552 and https://github.com/orgs/community/discussions/77614#discussioncomment-10598205 , I think this MIGHT be outdated.

**Comment by blaginin** [comment]
<img width="885" height="190" alt="Image" src="https://github.com/user-attachments/assets/ff7754e3-fcae-4ad9-a7b4-ad26cfff0c59" />

We ll need to make CI ready for the change and then open an email thread, will do that

**Comment by alamb** [comment]
> We ll need to make CI ready for the change and then open an email thread, will do that

I think the mailing list thread should basically say "we are discussing enabling the merge queue" and point at this ticket

**Comment by alamb** [comment]
Mailing List note was posted: https://lists.apache.org/thread/2kxhfgv10vhqt3y63lxwcjtlhxc55c4y ✅ 

> I'd like to test it first with a smaller repo (datafusion-sqlparser-rs) and move to the main one if everything works fine.

@iffyio, since this would impact you the most, being the most active maintainer of sqlparser-rs, would you be willing to try having merge queues enabled for that repo?

**Comment by iffyio** [comment]
@alamb yeah we can enable it on the sqlparser repo!

**Comment by blaginin** [comment]
Great infra team has given us temporary access to merge things bypassing the CI checks

<img width="818" height="109" alt="Image" src="https://github.com/user-attachments/assets/c1868119-0166-403a-88b0-554e65580bac" />

Hence we should be able to tests the queues in `sqlparser-rs` 🙂: https://github.com/apache/datafusion-sqlparser-rs/pull/2007


For transparency, here's the migration plan (https://issues.apache.org/jira/browse/INFRA-27154?page=com.atlassian.jira.plugin.system.issuetabpanels%3Aall-tabpanel)

<img width="1330" height="822" alt="Image" src="https://github.com/user-attachments/assets/21bf2b93-3b8a-4e26-9533-1a365981c613" />

**Comment by findepi** [comment]
From the mailing list

> This should hopefully help resolve some of the problems we have now:
> 
> - sometimes you want to approve a PR, but CI hasn't finished yet, so you have to wait up to 30 minutes before you can hit "merge"
> - the "merge" button may be available, but if the target branch was updated in the meantime, CI may then fail


The first should rarely be an issue in a distributed OSS projects. I've seen folks deliberately delaying merge by hours or days to give others a chance to chime in before the PR is merged.

How big of the problem is the second bullet?

Also, do we have any flaky tests? How will flakiness interact with merge queue?

not saying we should try it out. rather, trying to understand how important is the problem vs potential cost.

**Comment by crepererum** [comment]
> the "merge" button may be available, but if the target branch was updated in the meantime, CI may then fail

This "may" happen, but TBH I've rarely seen this. On the positive side as a maintainer I can just hit the queue button and walk away without having to worry about semantic/test conflicts (i.e. the ones not detected by GIT) on the target branch. That's also why the merge queue is rather efficient by default because it can test multiple entries of the queue, assuming that the previous entries will pass CI. If that assumption fails, entries will be retested with the failed entry removed.

What I am wondering though: what happens to commits in PR that are added/pushed AFTER the PR was added to the queue? IIRC that removes the PR from the queue (as one might expect, because we don't want people to sneak in changes) but I cannot find any first-party sources for that behavior.

> Also, do we have any flaky tests? How will flakiness interact with merge queue?

I'm not aware of any and I think we should be rather strict with them if they come up. Flaky tests just act like normal test failures.

**Comment by alamb** [comment]
@blaginin  has made a PR to SQLParser to test out how it works:
- https://github.com/apache/datafusion-sqlparser-rs/pull/2007

**Comment by blaginin** [comment]
Merge Queue is working in SQLParser 🥳


<img width="1039" height="389" alt="Image" src="https://github.com/user-attachments/assets/1aa296a9-9fff-4109-94f9-e9d76fbad5f6" />

<img width="1376" height="853" alt="Image" src="https://github.com/user-attachments/assets/ace57484-2c9f-490c-9933-8b92474fec7b" />

The plan is to test it for a few days, merge some PRs, and then switch to Datafusion if contributors will be happy with the experience

**Comment by blaginin** [comment]
Marked some steps as required in the main repo and asked ASF to enable merge queues for us here as well 🙂

**Comment by alamb** [comment]
so exciting!

**Comment by blaginin** [comment]
MQ just got enabled, I believe that should put the required checks in place https://github.com/apache/datafusion/pull/17538

**Comment by alamb** [comment]
Thank you for pushing this along @blaginin  (and debugging the issues that arise)

**Comment by blaginin** [comment]
Thank you! After https://github.com/apache/datafusion/pull/17629 is merged, I'll go through all the PRs and merge the once that got approved and we wanted to merge them but couldn't because of the check

**Comment by blaginin** [comment]
Also, I think in the future we need to create a [private?] sandbox project for experiments within the ASF GitHub org. I created my own, but apparently there are organization-level rules that are impossible to test if you just fork the repo :(

**Comment by alamb** [comment]
yeah, I agree testing stuff on github is tough

I think we can create our own datafusion repos within the apache org via some self service API -- how about `apache/datafusion-sandbox` ? 

https://infra.apache.org/version-control.html#create

**Comment by blaginin** [comment]
`datafusion-sandbox` is great! also i guess we should make it private not to confuse potential users

<img width="695" height="288" alt="Image" src="https://github.com/user-attachments/assets/fe9ceef6-4d23-44fc-94b9-8556d2120e46" />

I'm not sure I have permissions to create that, do you?

**Comment by alamb** [comment]
> I'm not sure I have permissions to create that, do you?

Yes, I think the PMC chair needs to request it. I will do so. 

I think it would be better if it is public with a clear README

**Comment by alamb** [comment]
https://github.com/apache/datafusion-sandbox is now available

**Comment by blaginin** [comment]
Thank you! My plan is to fork the main project there, but instead of readme put a big warning that you should never use this project, this is our CI sandbox

**Comment by blaginin** [comment]
Status: we're working on letting maintainers bypass branch protection rules if required steps fail (so cases like https://github.com/apache/datafusion/pull/17629 don't block us). Github support suggested some good ideas, and we're waiting for infra to set up a new permission group 🙏

**Comment by findepi** [comment]
How can I know what build jobs were performed by the merge queue?

I am intrigued by this:

<img width="547" height="129" alt="Image" src="https://github.com/user-attachments/assets/bc84e649-ad19-4153-83b5-4fd1b37b6fd6" />

- the PR https://github.com/apache/datafusion/pull/17872 was tested on CI 3 days ago
- there was a commit merged into main in the meantime
- yet, it was merged within a minute. I expected a full CI build to be performed by the merge queue, but I think it takes more time than that

**Comment by alamb** [comment]
> yet, it was merged within a minute. I expected a full CI build to be performed by the merge queue, but I think it takes more time than that

The idea is eventually to get to the point where the full CI build is performed as part of the Merge Queue 

However, we hit a problem with this (which effectively locked us out from being able to merge PRs)
- https://github.com/apache/datafusion/pull/17629

Ae are now in the process of trying to ensure we can bypass the checks when there is some need, but otherwise the checks will run as expected

**Comment by findepi** [comment]
Thanks for explaining! I noticed some references to a problem, but wasn't aware what the workaround actually entails.

**Comment by blaginin** [comment]
Status report for transparency: waiting for the infra team to enable the new setting

**Comment by Jefffrey** [comment]
I noticed that `Merge when ready` only waits for approvals instead of the actual CI checks to pass; is this something we are aware of/is expected?

**Comment by blaginin** [comment]
Correct, waiting for this Pr to merge: https://github.com/apache/infrastructure-actions/pull/312#issuecomment-3533995506
